File: Seafood Guide/Seafood Guide/View Controller/TVCDetailViewController.m
//
//  TVCDetailViewController.m
//  Seafood Guide
//
//  Created by Jon Brown on 9/2/14.
//  Copyright (c) 2014 Jon Brown Designs. All rights reserved.
//

#import "TVCDetailViewController.h"
#import "FishTVC.h"
#import <Social/Social.h>
#import "config.h"

@interface TVCDetailViewController ()

@end

@implementation TVCDetailViewController

@synthesize item;

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:UITableViewStyleGrouped];
    if (self) {
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [[self view] setBackgroundColor:[UIColor groupTableViewBackgroundColor]];
    
    self.title = [item name];
    
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:YES];
    
}
- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // #warning Potentially incomplete method implementation.
    // Return the number of sections.
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if(section==0) {
        return 3;
    } else {
        return 1;
    }
    
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{

    UITableViewCell *cell = nil;
    
    if (indexPath.section == 0) {

    if([indexPath row] == 0)
    {
        cell = [tableView dequeueReusableCellWithIdentifier:@"CellType1"];
        if(!cell)
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"CellType1"];
        [[cell textLabel] setText:[item name]];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    else if([indexPath row] == 1)
    {
        cell = [tableView dequeueReusableCellWithIdentifier:@"CellType2"];
        if(!cell)
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"CellType2"];
        [[cell textLabel] setText:[item region]];
        cell.textLabel.lineBreakMode = NSLineBreakByWordWrapping;
        cell.textLabel.numberOfLines = 0;
        cell.textLabel.font = [UIFont fontWithName:@"Helvetica" size:17.0];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    else if([indexPath row] == 2)
    {
        cell = [tableView dequeueReusableCellWithIdentifier:@"CellType3"];
        if(!cell)
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"CellType3"];
        [[cell textLabel] setText:[item desc]];
        cell.textLabel.lineBreakMode = NSLineBreakByWordWrapping;
        cell.textLabel.numberOfLines = 0;
        cell.textLabel.font = [UIFont fontWithName:@"Helvetica" size:17.0];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        
    }
    else if([indexPath row] == 3)
    {        
        cell = [tableView dequeueReusableCellWithIdentifier:@"CellType4"];
        if(!cell)
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"CellType4"];
        [[cell textLabel] setText:[item good]];
        cell.textLabel.lineBreakMode = NSLineBreakByWordWrapping;
        cell.textLabel.numberOfLines = 0;
        cell.textLabel.font = [UIFont fontWithName:@"Helvetica" size:17.0];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
        
        
    } else if (indexPath.section == 1) {
        
        if([indexPath row] == 0)
        {
        
        cell = [tableView dequeueReusableCellWithIdentifier:@"CellType6"];
        if(!cell)
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"CellType6"];
        [[cell textLabel] setText:@"Email to a friend"];
        
        }

    } 
    
    return cell;
}


#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    
    if (indexPath.section == 1) {
    
    if ([indexPath row] == 0) {
        
        [self sendEmail];
        
    }
    
    }
    
    
}



- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    if (indexPath.section == 0 && indexPath.row == 2) {
 
        NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle alloc] init];
        paragraphStyle.lineBreakMode = NSLineBreakByWordWrapping;
        
        NSDictionary *attributes = @{
                                     NSFontAttributeName:[UIFont systemFontOfSize:17],
                                     NSParagraphStyleAttributeName:paragraphStyle
                                     };
        
        CGRect textRect = [[item desc] boundingRectWithSize:CGSizeMake(290 - (10 * 2), 200000.0f)
                                               options:NSStringDrawingUsesLineFragmentOrigin
                                            attributes:attributes
                                               context:nil];
        
        float height = textRect.size.height;
        return height + (2 * 2);
        
    }
    return 44;
}


-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if (section == 0) {
        return @"Fish Information";
    } else if (section == 1) {
        return @"Share";
    }
    
    return nil;
}
-(void)viewDidAppear:(BOOL)animated
{
    //[super viewDidAppear:YES];
    [self.navigationItem setHidesBackButton:NO];
}

- (void)mailComposeController:(MFMailComposeViewController*)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError*)error
{
    switch (result)
    {
        case MFMailComposeResultCancelled:
            [ProgressHUD showError:@"Email Not Sent"];
            break;
        case MFMailComposeResultSaved:
            [ProgressHUD showSuccess:@"Email Saved!"];
            break;
        case MFMailComposeResultSent:
            [ProgressHUD showSuccess:@"Email Sent!"];
            break;
        case MFMailComposeResultFailed:
            [ProgressHUD showError:@"Email Not Sent"];
            break;
        default:
            [ProgressHUD showError:@"Email Not Sent"];
            break;
    }
    [self dismissViewControllerAnimated:YES completion:nil];
}


-(void)sendEmail {
    
    str = [item desc];
    str = [str substringToIndex: MIN(1165, [str length])];
    
    NSString *one = [item name];
    NSString *two = [item region];
    NSString *three = [item good];
    
    NSString* All = [NSString stringWithFormat:@"Fish Name: %@\n\nType: %@\n\n%@\n\nDescription:\n%@", one, two, three, str];
    
    
    
    
    if ([MFMailComposeViewController canSendMail])
    {
        MFMailComposeViewController *mailer = [[MFMailComposeViewController alloc] init];
        mailer.mailComposeDelegate = self;
        [mailer setSubject:[NSString stringWithFormat:@"Food & Water Watch - Check This Out: %@", one]];
        NSArray *toRecipients = [NSArray arrayWithObjects:[config getMail], nil];
        [mailer setToRecipients:toRecipients];
        NSString *emailBody = All;
        [mailer setMessageBody:emailBody isHTML:NO];
        [self presentViewController:mailer animated:YES completion:nil];
    }
    else
    {
        
        UIAlertController *alert = [UIAlertController
                                    alertControllerWithTitle:@"Failure"
                                    message:@"Your device doesn't support the composer sheet"
                                    preferredStyle:UIAlertControllerStyleAlert];
        
        
        UIAlertAction *noButton = [UIAlertAction
                                   actionWithTitle:@"OK"
                                   style:UIAlertActionStyleDefault
                                   handler:^(UIAlertAction * action) {
                                       //Handle no, thanks button
                                       [self.presentedViewController dismissViewControllerAnimated:NO completion:nil];
                                   }];
        
        [alert addAction:noButton];
        
        [self presentViewController:alert animated:YES completion:nil];
        
//        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Failure"
//                                                        message:@"Your device doesn't support the composer sheet"
//                                                       delegate:nil
//                                              cancelButtonTitle:@"OK"
//                                              otherButtonTitles:nil];
//        [alert show];
        
    }
}



-(UIButton*)createButtonWithFrame:(CGRect)frame andLabel:(NSString*)label
{
    
    UIButton *button = [[UIButton alloc] initWithFrame:frame];
    
    [button setTitleShadowColor:[UIColor blackColor] forState:UIControlStateNormal];
    button.titleLabel.shadowOffset = CGSizeMake(0, -1);
    button.titleLabel.font = [UIFont boldSystemFontOfSize:18];
    
    [button setTitle:label forState:UIControlStateNormal];
    [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    
    [[button layer] setBorderWidth:1.0f];
    [[button layer] setBorderColor:[UIColor whiteColor].CGColor];
    [[button layer] setMasksToBounds:YES];
    [[button layer] setCornerRadius:4.0]; //when radius is 0, the border is a rectangle
    [[button layer] setBorderWidth:1.0];
    
    return button;
}

-(void)closeView
{
    //[self dismissModalViewControllerAnimated:YES];
    [self dismissViewControllerAnimated:YES completion:nil];
    [self.navigationController popViewControllerAnimated:YES];
}

@end

File: Seafood Guide/Seafood Guide/View Controller/TVCDetailViewController.swift
//  Converted to Swift 5.1 by Swiftify v5.1.26565 - https://objectivec2swift.com/
//
//  TVCDetailViewController.swift
//  Seafood Guide
//
//  Created by Jon Brown on 9/2/14.
//  Copyright (c) 2014 Jon Brown Designs. All rights reserved.
//

import MessageUI
import ProgressHUD
import Social
import UIKit

class TVCDetailViewController: UITableViewController, UIWebViewDelegate, MFMailComposeViewControllerDelegate {
    var str = ""

    var item: Seafood?

    override init(style: UITableView.Style) {
        super.init(style: .grouped)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white

        title = item?.name
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)

    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

// MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        // #warning Potentially incomplete method implementation.
        // Return the number of sections.
        return 2
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return 3
        } else {
            return 1
        }

    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        var cell: UITableViewCell? = nil

        if indexPath.section == 0 {

            if indexPath.row == 0 {
                cell = tableView.dequeueReusableCell(withIdentifier: "CellType1")
                if cell == nil {
                    cell = UITableViewCell(style: .default, reuseIdentifier: "CellType1")
                }
                cell?.textLabel?.text = item?.name
                cell?.selectionStyle = .none
            } else if indexPath.row == 1 {
                cell = tableView.dequeueReusableCell(withIdentifier: "CellType2")
                if cell == nil {
                    cell = UITableViewCell(style: .default, reuseIdentifier: "CellType2")
                }
                cell?.textLabel?.text = item?.region
                cell?.textLabel?.lineBreakMode = .byWordWrapping
                cell?.textLabel?.numberOfLines = 0
                cell?.textLabel?.font = UIFont(name: "Helvetica", size: 17.0)
                cell?.selectionStyle = .none
            } else if indexPath.row == 2 {
                cell = tableView.dequeueReusableCell(withIdentifier: "CellType3")
                if cell == nil {
                    cell = UITableViewCell(style: .subtitle, reuseIdentifier: "CellType3")
                }
                cell?.textLabel?.text = item?.desc
                cell?.textLabel?.lineBreakMode = .byWordWrapping
                cell?.textLabel?.numberOfLines = 0
                cell?.textLabel?.font = UIFont(name: "Helvetica", size: 17.0)
                cell?.selectionStyle = .none
            } else if indexPath.row == 3 {
                cell = tableView.dequeueReusableCell(withIdentifier: "CellType4")
                if cell == nil {
                    cell = UITableViewCell(style: .subtitle, reuseIdentifier: "CellType4")
                }
                cell?.textLabel?.text = item?.good
                cell?.textLabel?.lineBreakMode = .byWordWrapping
                cell?.textLabel?.numberOfLines = 0
                cell?.textLabel?.font = UIFont(name: "Helvetica", size: 17.0)
                cell?.selectionStyle = .none
            }
        } else if indexPath.section == 1 {

            if indexPath.row == 0 {

                cell = tableView.dequeueReusableCell(withIdentifier: "CellType6")
                if cell == nil {
                    cell = UITableViewCell(style: .subtitle, reuseIdentifier: "CellType6")
                }
                cell?.textLabel?.text = "Email to a friend"
            }
        }

        return cell!
    }

// MARK: - Table view delegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {


        if indexPath.section == 1 {

            if indexPath.row == 0 {

                sendEmail()
            }
        }


    }

    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {

        if indexPath.section == 0 && indexPath.row == 2 {

            let paragraphStyle = NSMutableParagraphStyle()
            paragraphStyle.lineBreakMode = .byWordWrapping

            let attributes = [
                NSAttributedString.Key.font: UIFont.systemFont(ofSize: 17),
                NSAttributedString.Key.paragraphStyle: paragraphStyle
            ]

            let textRect = item?.desc?.boundingRect(with: CGSize(width: 290 - (10 * 2), height: 200000.0), options: .usesLineFragmentOrigin, attributes: attributes, context: nil)

            let height = Float(textRect?.size.height ?? 0.0)
            return CGFloat(height + (2 * 2))
        }
        return 44
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if section == 0 {
            return "Fish Information"
        } else if section == 1 {
            return "Share"
        }

        return nil
    }

    override func viewDidAppear(_ animated: Bool) {
        //[super viewDidAppear:YES];
        navigationItem.hidesBackButton = false
    }

    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        switch result {
            case .cancelled:
                ProgressHUD.failed("Email Not Sent")
            case .saved:
                ProgressHUD.succeed("Email Saved!")
            case .sent:
                ProgressHUD.succeed("Email Sent!")
            case .failed:
                ProgressHUD.failed("Email Not Sent")
            default:
                ProgressHUD.failed("Email Not Sent")
        }
        dismiss(animated: true)
    }

    func sendEmail() {

        str = item?.desc ?? ""
        str = (str as NSString).substring(to: min(1165, str.count))

        let one = item?.name
        let two = item?.region
        let three = item?.good

        let All = "Fish Name: \(one ?? "")\n\nType: \(two ?? "")\n\n\(three ?? "")\n\nDescription:\n\(str)"




        if MFMailComposeViewController.canSendMail() {
            let mailer = MFMailComposeViewController()
            mailer.mailComposeDelegate = self
            mailer.setSubject("Food & Water Watch - Check This Out: \(one ?? "")")
            mailer.setToRecipients([config.getMail()])
            let emailBody = All
            mailer.setMessageBody(emailBody, isHTML: false)
            present(mailer, animated: true)
        } else {

            let alert = UIAlertController(title: "Failure", message: "Your device doesn't support the composer sheet", preferredStyle: .alert)


            let noButton = UIAlertAction(title: "OK", style: .default, handler: { action in
                    //Handle no, thanks button
                    self.presentedViewController?.dismiss(animated: false)
                })

            alert.addAction(noButton)

            present(alert, animated: true)

            //        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Failure"
            //                                                        message:@"Your device doesn't support the composer sheet"
            //                                                       delegate:nil
            //                                              cancelButtonTitle:@"OK"
            //                                              otherButtonTitles:nil];
            //        [alert show];
        }
    }

    func createButton(withFrame frame: CGRect, andLabel label: String?) -> UIButton? {

        let button = UIButton(frame: frame)

        button.setTitleShadowColor(UIColor.black, for: .normal)
        button.titleLabel?.shadowOffset = CGSize(width: 0, height: -1)
        button.titleLabel?.font = UIFont.boldSystemFont(ofSize: 18)

        button.setTitle(label, for: .normal)
        button.setTitleColor(UIColor.white, for: .normal)

        button.layer.borderWidth = 1.0
        button.layer.borderColor = UIColor.white.cgColor
        button.layer.masksToBounds = true
        button.layer.cornerRadius = 4.0 //when radius is 0, the border is a rectangle
        button.layer.borderWidth = 1.0

        return button
    }

    func closeView() {
        //[self dismissModalViewControllerAnimated:YES];
        dismiss(animated: true)
        navigationController?.popViewController(animated: true)
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}
